### ISSUE TITLE
Edit Setup.ts to reflect database needs

### ISSUE THREAD

**Issue by Cioppolo14** [PalisadoesFoundation/talawa-api#1772]
**Describe the bug**
Currently, Talawa-api needs to have access to either a local database with a replica set or mongoDB atlas to use the retry writes option. Our setup.ts script does not support the user in ensure that one of these options is met. 

**Expected behavior**
Setup.ts needs to guide the user to make sure the database will work with retry rewrites. 

**Actual behavior**
Setup does not give the user this guidance. 

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

**Comment by adi790uu** [comment]
I would like to work on this.

**Comment by adi790uu** [comment]
@Cioppolo14 why we need to have access to local database with a replica set.

**Comment by Cioppolo14** [comment]
Its the developer's choice, either to run mongo atlas cluster or a local database converted to a replica set.   If the local database doesn't have a replica set, retry writes won't work.

**Comment by palisadoes** [comment]
It's going to have to be a mandatory requirement for non Atlas URLs for both the API database and API test database (which is created simultaneously)

**Comment by adi790uu** [comment]
ok @Cioppolo14 do you want me to write a script which automatically makes the local instance to be a replica set, something like this :

`
const { execSync } = require('child_process');

function executeCommand(command) {
  return execSync(command, { encoding: 'utf-8' });
}

const ports = [27017, 27018];
ports.forEach(port => {
  executeCommand(`mongod --port ${port} --dbpath /path/to/data/db${port}`);
});

executeCommand(`mongo --port 27017 --eval "rs.initiate({_id: 'myReplicaSet', members: [{ _id: 0, host: 'localhost:27017' },{ _id: 1, host: 'localhost:27018' }]})"`);

console.log(`mongodb://localhost:27017,localhost:27018/myDatabase?retryWrites=true`);
`

**Comment by adi790uu** [comment]
And in case of atlas Url make sure retryWrites=true parameter is present in the url?

**Comment by adi790uu** [comment]
@palisadoes can you review the above comments and give me your opinion.

**Comment by palisadoes** [comment]
We are awaiting feedback from @EshaanAgg on this topic

**Comment by adi790uu** [comment]
Okay @palisadoes

**Comment by EshaanAgg** [comment]
No need for a script. The user should be responsive for providing the correct URL. We should just inform the user of this difference through the setup script, and issue a warning if the user gives a local Mongo URL with retry writes enabled. The user should be able to bypass the warning if he choses to.

**Comment by adi790uu** [comment]
@EshaanAgg atleast we can attach retryWrites=true parameter to the Atlas url.

**Comment by palisadoes** [comment]
This solution assumes that the person installing the app will always be technically competent to make the decision. We need to default to values that will not cause future errors. 

It would be best to have the same behavior whether using a local DB vs the cloud. 

Either disabled in both cases or enabled with the appropriate warnings. The code should be written to expect one or the other configuration.

The installation script is to make installation foolproof

**Comment by adi790uu** [comment]
@palisadoes got it.

**Comment by palisadoes** [comment]
After doing more research on this topic,  @EshaanAgg is right.

Keep in mind that not all persons installing the app will be technically knowledgeable. The script must make installation as foolproof as possible.

**Comment by palisadoes** [comment]
@adi790uu @EshaanAgg @Cioppolo14 

1. The issue is caused by the latest implementation of event calendar entries.
2. If you are running on `localhost`, and you create an event you get this error. 
    1. We are trying to address this with this issue
    2. https://github.com/PalisadoesFoundation/talawa-api/issues/1583#issuecomment-1925469004
1. This wasn't the case previously. Therefore this may be a temporary fix. How best can we address this in the meantime?

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/65b33f1e-0423-452a-843f-f367f1e8eb0d)

**Comment by varshith257** [comment]
@palisadoes I encountered same while exploring talawa-admin. The potential reason for the recommendation to add retryWrites=false to MongoDB connection string could be a mismatch between the version of your MongoDB driver (mongoose) and the MongoDB server as most of us setup db locally in Atlas. 

For a temporary fix, adding retryWrites=false to end of connection string can be a workaround to address the immediate issue. However, we need the retryWrites for robust functionality of events. For future proof solutions we need make ensure compatibility between driver and server if need require to setup replicaSet as suggestions made following:  
https://stackoverflow.com/questions/58589631/mongoerror-this-mongodb-deployment-does-not-support-retryable-writes-please-ad

https://www.mongodb.com/docs/manual/core/retryable-writes/

https://mongoosejs.com/docs/compatibility.html

**Comment by adi790uu** [comment]
@palisadoes I have made the thing work based on this article: [https://dev.to/salim_angelo/how-to-convert-your-localhost-standalone-mongodb-cluster-into-a-replica-set-using-run-rs-npm-package-on-mac-os-54hd ](url), I created a replication set and I am not facing the error. 

<img width="1461" alt="Screenshot 2024-02-05 at 9 24 26 PM" src="https://github.com/PalisadoesFoundation/talawa-api/assets/132701661/ac14bb41-83e1-4714-b332-4384966d59a3">
<img width="1461" alt="Screenshot 2024-02-05 at 9 24 36 PM" src="https://github.com/PalisadoesFoundation/talawa-api/assets/132701661/a1c02d25-bef0-4349-84a5-599e52735a62">

**Comment by palisadoes** [comment]
This was triggered when we created recurring events. @meetulr the previous behavior needs to be restored as part of your revised recurring event issue.

- https://github.com/PalisadoesFoundation/talawa-api/issues/1583

I'm going to close this issue because we are working on a solution in the issue above.

**Comment by meetulr** [comment]
> This was triggered when we created recurring events. @meetulr the previous behavior needs to be restored as part of your revised recurring event issue.
> 
> * [Fix Recurring Events #1583](https://github.com/PalisadoesFoundation/talawa-api/issues/1583)
> 
> I'm going to close this issue because we are working on a solution in the issue above.

Do you mean we shouldn't use transactions? Replica set would be required for transactions.
If I removed transactions, things would work as before, but we wouldn't have that surety of everything passes or everything fails, as discussed [here](https://github.com/PalisadoesFoundation/talawa-api/pull/1658#issuecomment-1884968283).

**Comment by palisadoes** [comment]
@meetulr What would you suggest as the solution to make setup.ts work correctly for local and hosted Mongo instances?

Reopening

**Comment by meetulr** [comment]
@palisadoes I'm sorry I missed your comment.

Actually though, I was thinking about it.
I realized that rather than making a `replica set` mandatory, why not make `session` optional?

What I mean is that we could add a check for it in db.ts. After connecting to our mongoDB instance, we could check if it's a replica set or not, if it is, we'd start a session, if it's not, the session would not be initiated.

I've tried it locally, it's working.
